Ops and kernels for reduce_slice_ops

zasdfgbnm opened the pull request:
This is a PR for https://github.com/tensorflow/tensorflow/issues/7662

This PR implements `reduce_slice_sum`, `reduce_slice_prod`, `reduce_slice_max`, `reduce_slice_min`.

There are also tests for these new ops.

No gradients are implemented now, but I will start working on that after this PR is merged.

zasdfgbnm committed:
init impl done

zasdfgbnm committed:
improve doc, shape inference

zasdfgbnm committed:
add and pass shape inference

zasdfgbnm committed:
check bound

zasdfgbnm committed:
small fix

zasdfgbnm committed:
add all types of reduction

zasdfgbnm committed:
add cpu kernels

zasdfgbnm committed:
write python tests

zasdfgbnm committed:
make it able to build independent on GPU

zasdfgbnm committed:
fix problems and pass all tests

zasdfgbnm committed:
add to py

tensorflow-jenkins commented:
Can one of the admins verify this patch?

zasdfgbnm committed:
remove redundant code

girving commented:
This is implementable with current ops by doing a scan followed by a gather.  Is that not sufficient for your purposes?

drpngx commented:
I guess @zasdfgbnm has a point on the #7662. @aselle @girving what do you think?

girving commented:
@drpngx Yep, @zasdfgbnm and @aselle convinced me the op is reasonable, though I switched the reviewer to Andy.

caisq commented:
Gentle ping @aselle ?

aselle commented:
I'm sorry I'll take a thorough look today.

aselle commented:
Before merging this, can we think about if this will ever be useful to extend:
1. to higher order tensors than matrices. 
2. to axes that are not 0 and also if there are multiple axes?
If those things make sense and we can define them well. I'd like to make this function have a signature in the final form of that concept, so that we do not need to change the API to add that functionality.

It seems simple that we could define partial products of axis=1 right now where we sum over the columns rather than rows at least. This suggests we should have an axis parameter.

aselle commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -2365,4 +2365,132 @@ output_max: the computed max output.
+    .SetShapeFn(ParticalReductionShapeFn)
+    .Doc(R"doc(
+Dynamically compute the product over the first dimension of a tensor according
+to start and end indices specified at "index". For example if input is
```
<pre>
"index" -> `indices`
</pre>

aselle commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -2365,4 +2365,132 @@ output_max: the computed max output.
+    .Doc(R"doc(
+Dynamically compute the product over the first dimension of a tensor according
+to start and end indices specified at "index". For example if input is
+[[1,2,3],[10,20,30],[100,200,300],[10,10,10]] and index is [[0,1],[1,1],[0,2]],
```
Please use an example with different values in each column to make the example more illustrative.

aselle commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -2365,4 +2365,132 @@ output_max: the computed max output.
+    .Doc(R"doc(
+Dynamically compute the maximum over the first dimension of a tensor according
+to start and end indices specified at "index". For example if input is
+[[1,2,3],[10,20,30],[100,200,300],[10,10,10]] and index is [[0,1],[1,1],[0,2]],
```
It might be worth using pretty print to make the matrices multi line so they are more readable see array_ops.cc:884 "MatrixBandPart"

aselle commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -2365,4 +2365,132 @@ output_max: the computed max output.
+
+} // namespace
+
+REGISTER_OP("PartialSum")
```
Perhaps we should call this family of ops 
PartialReduceSum, PartialReduceProd, PartialReduceMin, PartialReduceMax so it is analogous to reduce_min, reduce_max, ...

zasdfgbnm replied:
Yes, I didn't think too much when naming this. Do you think naming them as RangedReduceSum, RangedReduceProd, RangedReduceMin, RangedReduceMax would be even better? The word "partial" is too general and the ops I'm adding can not do all types of "partial" reduction (for example adding all elements with odd index)

aselle replied:
It almost is a ReduceSliceSum becaue it's like you are taking a slice in one dimension and reducing along it? no?

zasdfgbnm replied:
Yes, "Slice" sounds better. Let's call it ReduceSliceSum for now before we can think of any better name.

aselle commented on tensorflow/python/ops/math_ops_test.py:
```diff
@@ -423,5 +423,72 @@ def testConsistent(self):
+      y_tf = math_ops.partial_min(x,indices).eval()
+      self.assertAllEqual(y_tf, result)
+
+  def testPartialReduceEmptyDataRows(self):
```
thanks for testing this case.

aselle commented on tensorflow/python/ops/math_ops_test.py:
```diff
@@ -423,5 +423,72 @@ def testConsistent(self):
+    with self.test_session(use_gpu=True):
+      y_tf = math_ops.partial_sum(x,indices).eval()
+      self.assertAllEqual(y_tf, result)
+
```
test scalars for indices and values?

zasdfgbnm replied:
Do you mean the case where "data" and "indices" are scalars? In my shape function, I explicitly forbid this case. See math_ops_test.cc: 493-495 in this PR

aselle reviewed (commented):
Not quite done reviewing yet...

aselle commented:
@tensorflow-jenkins test this please

zasdfgbnm commented:
Response to @aselle 
1. Higher order tensor is actually already supported now.
2. This is a great point, I think I will change the code to support this. It won't require too much work to do so.

zasdfgbnm committed:
improve doc

zasdfgbnm commented:
When implementing the reduction along any axis, I find it helpful to flat both inner and outer dims in the shape. So I create a new method `flat_inner_outer_dims` for class `Tensor` in a separate PR https://github.com/tensorflow/tensorflow/pull/9452

zasdfgbnm committed:
change ops signature

zasdfgbnm committed:
Merge remote-tracking branch 'upstream/master' into partial_reduces

drpngx commented:
Jenkins, test this please.

zasdfgbnm commented:
@drpngx this hasn't been done (actually still a lot to do). I just merge with master in order to use merged changes I made.

zasdfgbnm commented:
Before changing this to add support for user specified axis, I would like to add a `GetCuda3DLaunchConfig` to `tensorflow/core/util/cuda_kernel_helper.h`, in a separate PR. The newly added function will be used here.

drpngx commented:
Sounds good.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

rmlarsen commented:
@zasdfgbnm any update on your progress with GetCuda3DLaunchConfig?

zasdfgbnm commented:
@rmlarsen There is some part of the existing code that makes me not feeling well.
See https://github.com/tensorflow/tensorflow/issues/9937
Once everything is clear, I think it won't take too long to get a `GetCuda3DLaunchConfig`

rmlarsen commented:
@zasdfgbnm OK, thanks for the update! I'll mark this as stalled for now.

zasdfgbnm committed:
add Cuda{2D,3D}LaunchConfig that max occupancy

zasdfgbnm committed:
support axis

zasdfgbnm committed:
remove default val, check input<=0

zasdfgbnm committed:
modify tests to net api

zasdfgbnm committed:
fix some compilation err

zasdfgbnm committed:
Merge branch 'cuda_launch_cfg' into partial_reduces

zasdfgbnm committed:
Merge branch 'master' of upstream

zasdfgbnm committed:
fix Const in flat_inner_outer_dims

zasdfgbnm committed:
pass build

zasdfgbnm committed:
fix shape test

zasdfgbnm committed:
Merge branch 'unknown_shape_of_rank'

zasdfgbnm committed:
fix names in macros

zasdfgbnm committed:
fix test

zasdfgbnm committed:
specify reduceop by macro

zasdfgbnm committed:
clean code

zasdfgbnm committed:
further simplify code

zasdfgbnm committed:
misc fixes

zasdfgbnm committed:
add max size check

zasdfgbnm committed:
Merge branch 'cuda_launch_cfg'

zasdfgbnm committed:
fix typo

zasdfgbnm committed:
fix typo

zasdfgbnm committed:
Merge branch 'cuda_launch_cfg'

zasdfgbnm commented:
@aselle I think all my changes are done for this PR, please continue the review process. This op now allows users to specify the axis. I also added one test case to test when `axis != 0` and one test case to test when `rank > 2`. All these tests pass on my machine.

This PR relies on https://github.com/tensorflow/tensorflow/pull/10032 and https://github.com/tensorflow/tensorflow/pull/10068 that has not been merged into master branch yet. I manually merge them into this branch to make this work. So you don't have to worry about changes in `shape_inference.cc` and `cuda_kernel_helper.h` because they are reviewed separately. You may not want to merge this before these two dependencies are merged.

zasdfgbnm commented:
By the way, this PR include a fix in `tensor.h` related to const tensors

zasdfgbnm committed:
tests, docs, and related changes

zasdfgbnm committed:
build the test

zasdfgbnm committed:
buildify

shoyer commented:
Just to note, #8954 adds `tf.repeat`, which provides gradients for `partial_sum` (and vice-versa). See discussion at https://github.com/tensorflow/tensorflow/pull/8954#discussion_r118302537.

zasdfgbnm commented:
@shoyer The name of `partial_sum` has been suggested to change to `reduce_slice_sum`, see https://github.com/tensorflow/tensorflow/pull/9063#discussion_r113259298
The `tf.repeat` and `tf.reduce_slice_sum` are gradients of each other regardless of the API difference. But the current API of `tf.repeat` and `tf.reduce_slice_sum` makes it a little bit tricky to use directly:

- In order to use `tf.reduce_slice_sum` as gradient of `tf.repeat`, there must be another op to compute the `indices` that `tf.reduce_slice_sum` takes.
- In order to use `tf.repeat` as gradient of `tf.reduce_slice_sum`, there must be another op to reorder and merge some gradients, and compute the `repeats` argument that `tf.repeat` takes.

zasdfgbnm commented:
Currently, `indices` can only be a `n*2` matrix where the first column is the starting index and the second column is the ending index. It would be nice to extend the `indices` argument of `tf.reduce_slice_***` here to support a vector input. In the case of vector input, the second column of row `i` will be treated as the first column of row  `i+1`. For example:
```python
indices = [0,5,10,15]
``` 

is equivalent to

```python
indices = [ [0,5],
            [5,10],
            [10,15],
            [15,+infinity]]
```
This extension would make `tf.reduce_slice_***` more useful. And with this extension, the gradient of `tf.repeat` would become a very simple like
```python
tf.reduce_slice_sum(grad, tf.scan(lambda a, x: a + x, repeats)))
```

@zycdragonball I think in your pull request, you can submit a `tf.repeat` without implementation of gradient and wait for this PR to be merged. What do you think?

zasdfgbnm committed:
partially support vec indices

zasdfgbnm committed:
Merge branch 'master' of upstream

zasdfgbnm committed:
vec indices finish, not tested

zasdfgbnm committed:
pass cpu tests

zasdfgbnm committed:
Merge branch 'cuda_launch_cfg'

zasdfgbnm committed:
fix gpu functor

zasdfgbnm committed:
fix code style

zasdfgbnm committed:
update doc

zasdfgbnm committed:
fix code style

zasdfgbnm committed:
fix code style

zasdfgbnm commented:
The extend is done

zasdfgbnm committed:
Merge remote-tracking branch 'upstream/master'

zycdragonball commented:
@zasdfgbnm I'll leave the gradient of `tf.repeat` as it is now, and change it to `tf.reduce_slice_sum` once it is merged.

zasdfgbnm committed:
Merge upstream into partial_reduces

zasdfgbnm committed:
sync cuda_kernel_config change

drpngx commented:
Jenkins, test this please.

drpngx commented:
@zasdfgbnm this doesn't seem to build. Could you fix and push again?

zasdfgbnm committed:
Merge branch 'master' of upstream

zasdfgbnm committed:
buildify

zasdfgbnm committed:
typename->class in template template

typename in template template (See N4051: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4051.html)
is supported in gcc starting at 5(See: https://gcc.gnu.org/projects/cxx-status.html). So using typename instead
of class will cause the compilation fails at gcc-4.

zasdfgbnm commented:
@drpngx It should be fixed now.

drpngx commented:
Jenkins, test this please.

On Jun 27, 2017 4:46 AM, "Gao, Xiang" <notifications@github.com> wrote:

> @drpngx <https://github.com/drpngx> It should be fixed now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/9063#issuecomment-311333707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AT_SbQImWDEzLsY2HEtYiakCqB70dnOMks5sIOuagaJpZM4M3nG5>
> .
>

drpngx commented:
@aselle WDYT? If OK, then we can trigger an API review, or recommend it go to contrib.

aselle commented:
I'd like to put these into contrib for now. The reason being that we have repeat going into contrib, and this is needed for the gradients of repeat() and repeat() is needed for gradient here. Once they are both in contrib,  the gradients should be added, and once things have stabilized we should do api review and bring them in.

zasdfgbnm committed:
move to contrib

zasdfgbnm committed:
merge upstream

drpngx commented:
Jenkins, test this please.

zasdfgbnm commented:
Not done yet

zasdfgbnm committed:
some fixes

zasdfgbnm committed:
some fixes

zasdfgbnm committed:
some fixes

zasdfgbnm committed:
buildify

zasdfgbnm committed:
build cleanup

zasdfgbnm commented:
done now. should be working

_Sent from my OnePlus ONEPLUS A3000 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

vrv commented:
@tensorflow-jenkins test this please

zasdfgbnm commented:
seems to be unrelated fails?

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@zasdfgbnm It looks like you need to add your op to the Cmake build. 
https://ci.tensorflow.org/job/tensorflow-pr-win-cmake-py/3768/consoleFull

zasdfgbnm committed:
Merge branch 'master' of upstream

zasdfgbnm committed:
add cmake

zasdfgbnm commented:
@rmlarsen I added the cmake build rules. Hopefully it will work. I'm unable to test it on my machine because I'm running into this issue: https://github.com/tensorflow/tensorflow/issues/12018

drpngx commented:
Jenkins, test this please.

rmlarsen commented:
@zasdfgbnm Thanks!

rmlarsen commented:
@zasdfgbnm Thanks a lot for this nice contribution!

andrewharp commented on tensorflow/contrib/reduce_slice_ops/python/ops/reduce_slice_ops.py:
```diff
@@ -0,0 +1,30 @@
+_reduce_slice_ops = loader.load_op_library(
+    resource_loader.get_path_to_datafile("_reduce_slice_ops.so"))
+
+reduce_slice_sum = _reduce_slice_ops.reduce_slice_sum
```
This is giving me a problem with an internal build:
```
    reduce_slice_sum = _reduce_slice_ops.reduce_slice_sum
AttributeError: 'module' object has no attribute 'reduce_slice_sum'
```
Is there an extra step necessary somewhere to expose the ops completely?

zasdfgbnm replied:
Hi @andrewharp, I didn't see any problem on this line on my computer. But I do see another problem on somewhere else. I added some more commits to my branch. See the file changes in:
https://github.com/tensorflow/tensorflow/compare/master...zasdfgbnm:partial_reduces

Do you think this will fix the problem you are talking about?
